bika/health/upgrade/v01_01_003.py

```diff
@@ -31,7 +35,72 @@ def upgrade(tool):
     # Ensure health's skins have always priority over core's
     setup = portal.portal_setup
     setup.runImportStepFromProfile(profile, "skins")
+    setup.runImportStepFromProfile(profile, 'workflow')
+    setup.runImportStepFromProfile(profile, 'propertiestool')
+
+    add_doctor_action_for_client(portal)
+    ut.addIndexAndColumn('portal_catalog', 'allowedRolesAndUsers', 'FieldIndex')
+    ut.addIndex('portal_catalog', 'getPrimaryReferrerUID', 'FieldIndex')
+    ut.refreshCatalogs()
+
+    update_permissions_clients(portal)
 
     logger.info("{0} upgraded to version {1}".format(product, version))
 
     return True
+
+
+def add_doctor_action_for_client(portal):
+    # Add doctor action for client portal_type programmatically
+    client = portal.portal_types.getTypeInfo("Client")
+    for action in client.listActionInfos():
+        if action.get('id', None) == 'doctors':
+            logger.info("Already existing 'doctor' action for client portal_type")
+            return None
+    client.addAction(
+        id="doctors",
+        name="Doctors",
+        action="string:${object_url}/doctors",
+        permission=permissions.View,
+        category="object",
+        visible=True,
+        icon_expr="string:${portal_url}/images/doctor.png",
+        link_target="",
+        description="",
+        condition="")
+    logger.info("'doctor' action for client portal_type added")
+
+
+def update_permissions_clients(portal):
+    """
+    Maps and updates the permissions for clients.
+
+    :return: None
+    """
+    workflow_tool = api.get_tool("portal_workflow")
+    workflow = workflow_tool.getWorkflowById('bika_doctor_workflow')
+    catalog = api.get_tool('portal_catalog')
+    brains = catalog(portal_type='Doctor')
+    counter = 0
+    total = len(brains)
+    logger.info(
+        "Changing permissions for doctor objects: {0}".format(total))
+    for brain in brains:
+        if 'Client' not in brain.allowedRolesAndUsers:
```

Comment:
There is a thing I don't get here. You are increasing the counter each iteration of the loop but the message is logged when the counter is divisible by 100 and the current brain has `'Client'` in its allowed Roles & Users. I think this might result in missing some log messages if, for example, brain number 100 doesn't have `'Client'` in its allowed Roles & Users.   

I would either move the call to `logger.info()` out of the `if` statement or increase the counter only inside the conditional.

A part from that, I would maybe refactor the loop in @xispa 's style :joy: (which, from what I understood, mainly aims to reduce the amount of indentation levels):
If `counter` is increased only when a brain has `'Client'` in its allowed Roles & Users:
```python
for brain in brains:
    if 'Client' in brain.allowedRolesAndUsers:
        continue
    obj = api.get_object(brain)
    workflow.updateRoleMappingsFor(obj)
    obj.reindexObject()
    counter += 1
    if counter % 100 == 0:
        logger.info(
            "Changing permissions for doctor objects: " +
            "{0}/{1}".format(counter, total))
```
Otherwise:
```python
for brain in brains:
    counter += 1
    if counter % 100 == 0:
        logger.info(
            "Changing permissions for doctor objects: " +
            "{0}/{1}".format(counter, total))
    if 'Client' in brain.allowedRolesAndUsers:
        continue
    obj = api.get_object(brain)
    workflow.updateRoleMappingsFor(obj)
    obj.reindexObject()
```